File: citizen-ai-backend/src/models/WorkerTask.js
'use strict';
const mongoose = require('mongoose');

const workerTaskSchema = new mongoose.Schema({
  complaintId:      { type: mongoose.Schema.Types.ObjectId, ref: 'Complaint', required: true },
  assignedWorkerId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  assignedBy:       { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  assignedByName:   { type: String, default: 'Admin' },
  status:           { type: String, default: 'PENDING', enum: ['PENDING','ACCEPTED','REJECTED','WORK_STARTED','IN_PROGRESS','WORK_COMPLETED','COMPLETED'] },
  priority:         { type: String, default: 'NORMAL' },
  deadline:         { type: Date, default: null },
  workerNotes:      { type: String, default: '' },
  rejectReason:     { type: String, default: null },
  progressPercentage: { type: Number, default: 0 },
  acceptedAt:       { type: Date, default: null },
  startedAt:        { type: Date, default: null },
  completedAt:      { type: Date, default: null },
}, { timestamps: true });

workerTaskSchema.index({ assignedWorkerId: 1 });
workerTaskSchema.index({ complaintId: 1 });
workerTaskSchema.index({ status: 1 });

workerTaskSchema.methods.toDTO = function (complaint) {
  const iso = (d) => d ? new Date(d).toISOString() : null;
  const c = complaint || this._complaint;
  return {
    id: this._id.toString(),
    complaintId: this.complaintId.toString(),
    issueType: c ? c.issueType : '',
    category: c ? c.category : '',
    address: c ? c.address : '',
    latitude: c ? c.latitude : 0,
    longitude: c ? c.longitude : 0,
    priority: this.priority || (c ? c.priority : 'NORMAL'),
    status: this.status,
    beforeImageUrl: c ? c.imageUrl : null,
    afterImageUrl: c ? c.afterImageUrl : null,
    citizenDescription: c ? c.description : '',
    workerNotes: this.workerNotes,
    assignedAt: iso(this.createdAt),
    acceptedAt: iso(this.acceptedAt),
    startedAt: iso(this.startedAt),
    completedAt: iso(this.completedAt),
    deadline: iso(this.deadline),
    progressPercentage: this.progressPercentage,
    assignedBy: this.assignedBy ? this.assignedBy.toString() : null,
    assignedByName: this.assignedByName,
    distanceKm: null,
  };
};

module.exports = mongoose.model('WorkerTask', workerTaskSchema);

File: CitizenAI/app/src/main/java/com/citizenai/app/domain/model/WorkerTask.kt
package com.citizenai.app.domain.model

import java.time.Instant

/**
 * WorkerTask — a complaint assignment visible to a Worker.
 * Mirrors the Complaint model but includes worker-specific fields.
 *
 * Status flow: PENDING → ACCEPTED / REJECTED → IN_PROGRESS → COMPLETED / REOPENED
 */
data class WorkerTask(
    val id: String,
    val complaintId: String,       // e.g. "CIV-1024"
    val issueType: String,
    val category: IssueCategory,
    val address: String,
    val latitude: Double,
    val longitude: Double,
    val priority: Priority,
    val status: TaskStatus,
    val beforeImageUrl: String?,
    val afterImageUrl: String?,
    val citizenDescription: String,
    val workerNotes: String?,
    val assignedAt: Instant,
    val acceptedAt: Instant? = null,
    val startedAt: Instant?,
    val completedAt: Instant?,
    val deadline: Instant? = null,
    val progressPercentage: Int = 0,
    val assignedBy: String? = null,
    val assignedByName: String? = null,
    val distanceKm: Double?
)

enum class TaskStatus {
    PENDING,       // Assigned but not yet accepted by worker
    ACCEPTED,      // Worker accepted
    REJECTED,      // Worker rejected (reassignment needed)
    IN_PROGRESS,   // Worker started work
    COMPLETED,     // Worker submitted completion
    REOPENED;      // Citizen rejected completion — back to active

    fun displayLabel(): String = when (this) {
        PENDING    -> "Pending Acceptance"
        ACCEPTED   -> "Accepted"
        REJECTED   -> "Rejected"
        IN_PROGRESS -> "In Progress"
        COMPLETED  -> "Completed"
        REOPENED   -> "Reopened"
    }
}
